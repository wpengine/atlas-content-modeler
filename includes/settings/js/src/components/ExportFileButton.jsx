import React from "react";
import { __ } from "@wordpress/i18n";
import { showSuccess, showError } from "../toasts";
import { DarkButton } from "../../../../shared-assets/js/components/Buttons";

export default function ExportFileButton({
	fileData,
	fileTitle,
	successMessage,
	errorMessage,
	buttonTitle,
	callbackFn,
	buttonClasses,
	fileType,
}) {
	/**
	 * Get data and type ready for export
	 * @param data
	 * @returns {{data, type: string}|{data: string, type: string}}
	 */
	function processDataForExport(data) {
		switch (fileType) {
			case "json":
				return {
					data: JSON.stringify(data),
					type: "application/json",
				};
			default:
				return { data: data, type: "text/plain" };
		}
	}

	/**
	 * Export final file - json
	 * @param data
	 */
	function exportFile(data) {
		const readyData = processDataForExport(data);
		const blob = new Blob([readyData.data], {
			type: readyData.type,
		});
		const url = URL.createObjectURL(blob);
		const link = document.createElement("a");
		link.download = fileTitle;
		link.href = url;
		link.click();

		showSuccess(
			__(
				successMessage || "The export was successful.",
				"atlas-content-modeler"
			)
		);
	}

	/**
	 * Export click handler for generating models .json file
	 * @param event
	 */
	function exportClickHandler(event) {
		event.preventDefault();

		if (callbackFn && !fileData) {
			callbackFn()
				.then((res) => {
					if (res) {
						exportFile(res);
					} else {
						showError(
							__(
								"There is no data to export.",
								"atlas-content-modeler"
							)
						);
					}
				})
				.catch(() => {
					showError(
						__(
							errorMessage ||
								"There was an error exporting the data.",
							"atlas-content-modeler"
						)
					);
				});
		} else if (fileData) {
			exportFile(fileData);
		}
	}

	return (
		<DarkButton
			data-testid="export-models-button"
			onClick={(event) => exportClickHandler(event)}
		>
			{__(buttonTitle || "Export", "atlas-content-modeler")}
		</DarkButton>
	);
}
