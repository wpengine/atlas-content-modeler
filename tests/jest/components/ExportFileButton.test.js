import ExportFileButton from "../../../includes/settings/js/src/components/ExportFileButton";
import ShallowRenderer from "react-test-renderer/shallow";
import React from "react";

describe("ExportFileButton tests", () => {
	const fileData = {};
	const renderer = new ShallowRenderer();
	renderer.render(<ExportFileButton fileData={fileData} />);

	const tree = renderer.getRenderOutput();

	it("Renders a matching snapshot", () => {
		expect(tree).toMatchSnapshot();
	});
});
