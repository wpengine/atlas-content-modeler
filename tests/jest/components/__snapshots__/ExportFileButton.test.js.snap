// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportFileButton tests Renders a matching snapshot 1`] = `
<button
  className="button dark-blue button-primary link-button"
  onClick={[Function]}
>
  Export
</button>
`;
